rfd/0193-access-control-decision-api.md

```diff
@@ -0,0 +1,345 @@
+and to ensure that it is as difficult as possible to accidentally misuse it either by misunderstanding or oversight.
+
+### API
+
+The PDP/PEP boundary will be represented as a gRPC API, with methods for the different kinds of decisions. Because teleport
+as a whole is an integrated PDP/PEP, our decisions are often much more nuanced than a simple allow/deny. This is one of
+teleport's greatest strengths, as this allows us to provide very granular controls over actions. It also means that many
+of the decisions we make are very verbose. Take ssh access for example: an 'allow' decision for ssh access must include
+parameters for agent forwarding, port forwarding, concurrent connection limits, BPF events, expired cert disconnect, locking,
+file copying, etc. For this reason we will eschew any attempt at creating a "unified" method for decisions in favor of
+implementing custom methods and types intended to meet the specific needs of specific decisions.
+
+In order to ensure that the PDP API remains simple and usable we will establish a set of conventions for how decision methods
+should be implemented and how types should be structured:
+
+- All "allow" decisions will be structured as conditional allows, pending the application of limits/parameters. All allow decisions
+will be communicated by a type of the form `<Action>Permit`, which will encode all limits/parameters to be applied by the PEP. Much of
+the agent-side teleport logic that previously would have worked with some combination of certificate identity and role set will instead
+take the appropriate permit structure as input.
+
+- Decision Service methods will conventionally take the form `rpc Evaluate<Action>(Evaluate<Action>Request) returns (Evaluate<Action>Response)`.
+For example, the method for evaluating server access would read `rpc EvaluateSSHAccess(EvaluateSSHAccessRequest) returns (EvaluateSSHAccessResponse)`.
+
+- Response objects will contain a top-level `Decision` oneof field with `Permit` and `Denial` variants. The underlying permit/denial
+types will conventionally be named `<Action>Permit` and `<Action>Denial`.
+
+- Common `RequestMetadata`, `PermitMetadata`, and `DenialMetadata` types will be provided and should be included in all requests, permits,
+and denials. These types will contain common fields like teleport version of PDP/PEP, dry run flag, etc.
+
+- We will not generally try to write polymorphic abstractions over different request/decision/permit/denial types, but helpers will be
+provided for basic permit/denial control-flow. For example, a `decision.Unwrap` function will be provided that converts decision responses
+into the form `(Permit, error)` where `error` is a `trace.AccessDeniedError` built from the contents of the denial.
+
+- A standard `Resource` reference types with fields like  `kind` and `name` will be provided and should be prefered in
```

Comment:
What about passing down the kubernetes resources as part of the permit? I would rather not have to deal with multiple permits or do the combinatory explosion at the control plane.

Reply:
There are cases where the `kubernetes_groups` are deeply tied together with `kubernetes_resources`.

For instance, when you run `kubectl delete pods --all --namespace <namespace>`, a request is made to `DELETE /api/v1/namespaces/<namespace>/pods`.

Given that `kubernetes_resources` might restrict the resources you can access, Teleport transforms the request into:

- `GET /api/v1/namespaces/<namespace>/pods`
- It filters the pods the user doesn't have access
- For each pod, it computes the `kubernetes_groups` the user is allowed to use. This requires ${kubernetes}\textunderscore{labels} \cap {kubernetes}\textunderscore{resources}$ for each POD name, namespace and verb. For this, the PEP must know the possible set of space $\mathbb{D}({kubernetes}\textunderscore{resources}) \rightarrow {kubernetes}\textunderscore{groups}$
- Finally, for each pod it executes the request `DELETE /api/v1/namespaces/<namespace>/pods/<pod_name>`

Because of this, it requires a more complex representation of permits.

Reply:
I'm convinced that k8s complexity should warrant the change from single to multi-permit per request. What alternatives do you see that would allow us to keep a single permit?

Reply:
We can update the `K8SPermit` structure as follows:

```proto
message K8SPermit {
    repeated KubeResourcesGroupMappings mappings = 1;

    // Note: No kubernetes_users or kubernetes_groups directly under this message
}

message KubeResourcesGroupMappings {
    repeated string kubernetes_groups = 1;
    repeated string kubernetes_users = 2;
    repeated KubernetesResources kubernetes_resources = 3;
}

message KubernetesResources {
    string kind = 1;
    string api = 2;
    string name = 3;
    optional string namespace = 4;
    repeated string verbs = 5;
}
```

This will become more complex to manage if we introduce other restrictions like `k8s_resources` (not planned) because requires changing the messages and possibly introducing more fields while in permits it would be "easier"

Reply:
After thinking a bit more about this problem, I think we will need to have multiple permits (allow/deny) or duplicate the mappings. This is necessary because users might have denied `kubernetes_groups` for specific `kubernetes_resources`. 

Like:

```
allow:
    kubernetes_groups: ["write","read"]
    kubernetes_resources:
       - kind: *
         name: *
         namespace: *
         verbs: *
deny:
    kubernetes_groups: ["write"]
    kubernetes_resources:
       - kind: namespace
         name: kube-system
         verbs: *
```

This would require us to change the messages to allow multi permits or duplicate mappings like bellow 

```
message K8SPermit {
    repeated KubeResourcesGroupMappings allow_mappings = 1;
    repeated KubeResourcesGroupMappings deny_mappings = 1;
    // Note: No kubernetes_users or kubernetes_groups directly under this message
}
```